README

This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                  Jörg Wendel <vdr@jwendel.de>

Project's homepage:          URL


This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
See the file COPYING for more information.

Description:

Make shure that vdr user has sufficiant priveleges to Access the USB serial port.
e.g. to ensure use an udev.rule like this:
KERNEL=="ttyUSB?", OWNER="vdr", ATTRS{product}=="SEDU Board USB", SYMLINK+="ttySEDU"

Adjust the seduconf.pl cofig script to generate a suiteable seduatmo.conf according to your configuration and place it in plugin cond dir.

Blacklevel = 5         -> brightness for blacklevel
DetectCineBars = 2     -> 0- horizontal; 1 - vertical; 2 - both
FixedColorBlue = 0     -> B level of fixek color (0-255)
FixedColorGreen = 101  -> G level of fixek color (0-255)
FixedColorRed = 111    -> R level of fixek color (0-255)
Frequence = 25         -> frequency
Gamma = 14             -> Gamma correction (divided by 10)
HorizontalDepht = 2    -> Number of PIX extend into screen on left/right, 0 for the most outer one only
VerticalDepht = 1      -> Number of PIX extend into screen at top/bot, 0 for the most outer one only
LevelBlue = 70         -> white calibration level blue(0-255)
LevelGreen = 100       -> white calibration level green(0-255)
LevelRed = 57          -> white calibration level red(0-255)
LogLevel = 0           -> log Level (0-2)
SeduMode = 0           -> Protocol: 0 – miniDMX (only A2/512Byte), 1 - tpm2
SeduRgbOrder = BGR     -> ;)
ShowMainmenu = 1       -> ;)
Threshold = 17         -> level for black level detection, dark led until all LED below this threshold
ViewMode = 0           -> 0 - atmo; 1 - fixed color; 2 - off

Blacklevel = 5         -> Helligkeitslevel bei Schwarzem Bild
DetectCineBars = 2     -> 0- horizontal; 1 - vertical; 2 - both
FixedColorBlue = 0     -> B Wert für Fixed Color (0-255)
FixedColorGreen = 101  -> G Wert für Fixed Color (0-255)
FixedColorRed = 111    -> R Wert für Fixed Color (0-255)
Frequence = 25         -> refresh Frequenz 
Gamma = 14             -> Gamma Korrektur (wird durch 10 geteilt)
HorizontalDepht = 2    -> Anzahl der Pixel die seitlich in das Bild hinein herangezogen werden, 0 für nur das äußere
VerticalDepth = 1      -> Anzahl der Pixel die oben und unten in das Bild hineinragen , 0 für nur das äußere
LevelBlue = 70         -> Weißabgleich B (0-255)
LevelGreen = 100       -> Weißabgleich G (0-255)
LevelRed = 57          -> Weißabgleich R (0-255)
LogLevel = 0           -> log Level (0-2)
SeduMode = 0           -> Protokoll: 0 – miniDMX (nur A2/512Byte), 1 - tpm2
SeduRgbOrder = BGR     -> ;)
ShowMainmenu = 1       -> ;)
Threshold = 17         -> Schwellwert für Schwarzerkennung, liegen alle drei Farben eines Pixel darunter bleibt die LED aus
ViewMode = 0           -> 0 - atmo; 1 - fixed color; 2 - off


HISTORY
VDR Plugin 'seduatmo' Revision History
--------------------------------------

2016-01-03: Version 0.0.4
      added handling of rpg order per led

2015-10-11: Version 0.0.3
      Fixed menu handling

2012-11-26: Version 0.0.2

2012-11-07: Version 0.0.1
  - Initial revision.


setup.h
/*
 * stup.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: seduatmo.c,v 1.64 2012/11/28 06:29:24 wendel Exp $
 */

//***************************************************************************
// Setup
//***************************************************************************

class cSeduSetup : public cMenuSetupPage, public cSeduService
{
   public:

      cSeduSetup();

      static const char* seduRGBOrders[];
      static int toOrderIndex(const char* order);

   protected:

      virtual void Setup();
      virtual eOSState ProcessKey(eKeys Key);
      virtual void Store();

      const char* seduModes[smCount];
      const char* cineBars[cbCount];
      int rgbOrderIndex;

      cSeduConfig data;
};


config.c
/*
 * seduthred.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: config.c,v 1.27 2012/11/28 06:29:24 wendel Exp $
 */

#include <string.h>

#include "config.h"
#include "common.h"

//***************************************************************************
// Global Configuration
//***************************************************************************

cSeduConfig cfg;

//***************************************************************************
// cConfigData
//***************************************************************************

cSeduConfig::cSeduConfig()
{
   // to be configured

   frequence = 25;
   threshold = 17;
   average = 10;
   adjRed = 57;
   adjGreen = 100;
   adjBlue = 70;
   gamma = 14;
   xDeep = 2;
   yDeep = 1;
   black = 0;
   detectCineBars = cbBoth;

   seduMode = smMiniDMX;
   strcpy(seduRGBOrder, "BGR");
   loglevel = 0;

   showMainmenu = yes;
   viewMode = vmAtmo;
   effectSpeed = 500;

   fixedR = 111;
   fixedG = 101;
   fixedB = 0;


   // calculated

   leds = 0;
   ledCount = 0;
   grabWidth = na;
   grabHeight = na;
}

cSeduConfig::~cSeduConfig()
{
   if (leds) delete leds;
}

//***************************************************************************
// Create Leds from config
//***************************************************************************

cSeduConfig::cLed* cSeduConfig::createLeds(cLedConfs* conf)
{
   int seq = 0;
   delete leds; 
   
   grabWidth = 0;
   grabHeight = 0;
   
   ledCount = conf->Count();
   leds = new cLed[ledCount];
   memset(leds, 0, ledCount*sizeof(cLed));
   
   for (cLedConf* l = conf->First(); l; l = conf->Next(l))
   {
      tell(1, "led%d (%d) %d/%d  %d/%d", seq, l->Pos(),
           l->X(), l->Y(), 
           l->ToX(), l->ToY());
      
      if (l->isValid())
      {
         // calc size of led matrix 
         
         if (grabWidth < l->X())
            grabWidth = l->X();
         if (grabWidth < l->ToX())
            grabWidth = l->ToX();
         if (grabHeight < l->Y())
            grabHeight = l->Y();
         if (grabHeight < l->ToY())
            grabHeight = l->ToY();
         
         leds[seq].lp  = (LedPosition)l->Pos();
         leds[seq].x   = l->X();
         leds[seq].y   = l->Y(); 
         leds[seq].toX = l->ToX();
         leds[seq].toY = l->ToY();
         sprintf(leds[seq].rgbOrder, "%.3s", l->RgbOrder());
         
         seq++;
      }
   }

   ledCount = seq;
   grabWidth++;
   grabHeight++;

   return leds;
}


ledsconf.c
/*
 * ledsconf.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: ledsconf.c,v 1.18 2012/11/20 19:04:09 wendel Exp $
 */

#include <ctype.h>

#include "common.h"
#include "config.h"
#include "ledsconf.h"
#include "setup.h"

//***************************************************************************
// cLedConf
//***************************************************************************
//***************************************************************************
// Object
//***************************************************************************

cLedConf::cLedConf() 
{ 
   x = na;
   toX = na;
   y = na;
   toY = na;
   lp = na;
}

//***************************************************************************
// Parse like like "led bot 0-1 14-17"
//
//   # left
//   led left 0  0
//   led left 0  1
//   led left 0  2
//   led left 0  3
//   led left 0  4
//   led left 0  5
//   led left 0  6
//   led left 0  7
//   
//***************************************************************************

bool cLedConf::Parse(const char* s)
{
   const char* p = s;

   p = skipWs(p);

   // check keyword

   if (strncasecmp(p, "led ", 4) != 0)
      return false;

   p += 4;
   skipWs(p);

   // LED Position

   if (strncasecmp(p, "top ", 4) == 0)
      lp = lpTop;
   else if (strncasecmp(p, "left ", 5) == 0)
      lp = lpLeft;
   else if (strncasecmp(p, "bot ", 4) == 0)
      lp = lpBottom;
   else if (strncasecmp(p, "bottom ", 7) == 0)
      lp = lpBottom;
   else if (strncasecmp(p, "right ", 6) == 0)
      lp = lpRight;
   else
      return error("Missing location {top,left,bot(tom),right}");
   
   // skip to delemiter
   
   while (*p && *p != ' ' && *p != '\t')
      p++;

   // check

   if (!*p)
      return false;

   skipWs(p);

   // parse X 

   if (!parseRange(p, x, toX))
      return false;

   // parse Y
   
   if (!parseRange(p, y, toY))
      return false;
   
   // parse optinal RGB order

   if (!parseRgbOrder(p, rgbOrder))
      return false;
   
   return true;
}

//***************************************************************************
// Parse Range like "12-26"
//***************************************************************************

bool cLedConf::parseRange(const char*& p, int& from, int& to)
{
   p = skipWs(p);

   if (!isdigit(*p))
      return false;

   from = to = strtol(p, (char**)&p, 0);

   p = skipWs(p);

   if (*p != '-')
      return true;

   p++;

   if (!isdigit(*p))
      return false;

   to = strtol(p, (char**)&p, 0);

   return true;
}

//***************************************************************************
// Parse RGB Order
//***************************************************************************

bool cLedConf::parseRgbOrder(const char*& p, char* rgbOrder)
{
   int size ;
   p = skipWs(p);

   size = strlen(p);

   strcpy(rgbOrder, cfg.seduRGBOrder);  // the defaul

   if (size == 0)
      return true;

   if (cSeduSetup::toOrderIndex(p) == na)
      return error("Invalid RGB order '%s' configured", p);

   strcpy(rgbOrder, p);

   return true;
}

//***************************************************************************
// Skip Whitespaces
//***************************************************************************

const char* cLedConf::skipWs(const char* p)
{
   while (*p && (*p == ' ' || *p == '\t'))
      p++;

   return p;
}


ledsconf.h
/*
 * ledsconf.h: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: ledsconf.h,v 1.9 2012/11/20 14:32:28 wendel Exp $
 */

#ifndef _LED_CONF_H_
#define _LED_CONF_H_

#include <vdr/plugin.h>
#include <vdr/tools.h>

#include "seduservice.h"

//***************************************************************************
// cLedConf
//***************************************************************************

class cLedConf : public cListObject, public cSeduService
{
   public:
      
      cLedConf();
      
      bool Parse(const char* s);

      int X()                { return x; }
      int ToX()              { return toX; }
      int Y()                { return y; }
      int ToY()              { return toY; }
      int Pos()              { return lp; }
      int isValid()          { return x > na && y > na && lp > na; }
      const char* RgbOrder() { return rgbOrder; }

   private:

      bool parseRange(const char*& s, int& from, int& to);
      bool parseRgbOrder(const char*& p, char* rgbOrder);
      const char* skipWs(const char* p);

      int x;
      int toX;
      int y;
      int toY;
      int lp;
      char rgbOrder[3+TB];
};

//***************************************************************************
// cLedConfs
//***************************************************************************

class cLedConfs : public cConfig<cLedConf> 
{

};

extern cLedConfs ledConfs;

#endif // _LED_CONF_H_


seduthread.h
/*
 * seduthread.h: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: seduthread.h,v 1.64 2012/11/28 06:29:24 wendel Exp $
 */

#include <termios.h>
#include <queue>

#include <vdr/thread.h>

#include "common.h"
#include "config.h"

class PixQueue
{
   public:

      PixQueue()            { clear(); }
      virtual ~PixQueue()   { clear(); }
      void clear()          { while (!pQueue.empty()) pQueue.pop(); r=g=b=0; }
      int getCount()        { return pQueue.size(); }

      void push(Pixel* p) 
      { 
         pQueue.push(*p);
 
         r += p->r;
         g += p->g;
         b += p->b;        

         // remove old entrys from queue and sum

         while (pQueue.size() > cfg.average)
         {
            r -= pQueue.front().r;
            g -= pQueue.front().g;
            b -= pQueue.front().b;
            
            pQueue.pop();
         }
      }

      void getPixel(Pixel* p)
      {
         p->r = r / pQueue.size();
         p->g = g / pQueue.size();
         p->b = b / pQueue.size();
      }
      
   protected:

      std::queue <Pixel> pQueue;
      unsigned int r;
      unsigned int g;
      unsigned int b;
};

class PixSum
{
   public:

      PixSum()       { clear(); }

      void clear()   { r=g=b=count=0;  }
      int getCount() { return count; }

      void add(Pixel* p)
      {
         r += p->r;
         g += p->g;
         b += p->b;
         count++;
      }

      void getAvg(Pixel* p)
      {
         p->r = r / count;
         p->g = g / count;
         p->b = b / count;
      }

   protected:
      
      unsigned int r;
      unsigned int g;
      unsigned int b;
      int count;
};

//***************************************************************************
// cSeduLine - Serial Communication
//***************************************************************************

class cSeduLine : public cSeduService
{
   public:

      cSeduLine();
      ~cSeduLine()  { close(); }

      int open();
      int close();
      int isOpen()  { return fd != na; }

      int checkLine();
      int read();

      int write(unsigned char b);

      int writeStartSeq();
      int writeEndSeq();
      int writePix(int ledIdx, Pixel* p);
      int writeColor(int ledIdx, Pixel* p, int index);

      // set ..

      void setMode(SeduMode aMode, int channels = 0);

   private:

      int detect();
   
      // data

      int dataBytesSend;

      int mode;
      int dataBytes;
      int byteMode;
      int byteStart;
      int byteEnd;

      char* deviceName;
      int fd;
      struct termios oldtio;
};

//***************************************************************************
// SEDU Thread
//***************************************************************************

class cSeduThread : public cThread, public cSeduService
{
   public:

      cSeduThread();
      ~cSeduThread();

      int isRunning() { return Running(); }

      // interface

      void Stop();
     
   private:

      void Action(void);

      int grabImage();
      int detectCineBars();
      int putData();

      void threshold(Pixel* p);
      void whiteAdj(Pixel* p);
      void gammaAdj(Pixel* p);
      void merge(Pixel* p1, Pixel* p2, int level = 100);

      int getPixel(int ledIdx, Pixel* pixel);
      int isBlack(Pixel* p) { return p->r < cfg.threshold && p->g < cfg.threshold &&  p->b < cfg.threshold; }

      // data

      cSeduLine sedu;

      cMutex mutex;
      cCondVar waitCondition;
      int loopActive;

      Pixel* image;
      PixQueue* pixAverage;
      int cineBarsHor;
      int cineBarsVer;
      int imageSize;
      int imageWidth;
      int imageHeight;

      Pixel getRainbowColor();
      Pixel getColorWheel(int moving = 0, int led = 0);
};


seduservice.h
/*
 * seduservice.h: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: seduservice.h,v 1.3 2012/11/22 18:02:59 wendel Exp $
 */

#ifndef __SEDU_SERVICE_H
#define __SEDU_SERVICE_H

#include "common.h"

//***************************************************************************
// Pixel - format as provided by softhddevice 
//***************************************************************************

struct Pixel
{
   unsigned char b;
   unsigned char g;
   unsigned char r;
   unsigned char a;
};

//***************************************************************************
// Sedu Service
//***************************************************************************

class cSeduService
{
   public:

      enum Cinebars
      {
         cbHorizontal,
         cbVertical,
         cbBoth,
         cbCount
      };

      enum SeduMode
      {
         smMiniDMX,
         smTpm2,
		 smMiniDMX_A1,
		 smMiniDMX_B0,
         smCount
      };

      enum ViewMode
      {
         vmAtmo,
         vmFixedCol,
         vmRainbow,
         vmColorWheel,
         vmColorWheelStatic,
         vmBlack,
         vmDetached,
         vmCount
      };

      enum LedPosition
      {
         lpTop,
         lpLeft,
         lpBottom,
         lpRight
      };

      struct cLed
      {
         int x;
         int toX;
         int y;
         int toY;
         LedPosition lp;
         char rgbOrder[3+TB];
      };

      // static

      void rgb2hsv(int r, int g, int b, double* h, double* s, double* v);
      Pixel hsv2rgb(int h, double s, double v);

      static const char* toName(ViewMode vm);
      static const char* viewModes[];
};

#endif // __SEDU_SERVICE_H


seduatmo.c
/*
 * seduatmo.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: seduatmo.c,v 1.64 2012/11/28 06:29:24 wendel Exp $
 */

#include <vdr/plugin.h>

#include "seduthread.h"
#include "config.h"
#include "ledsconf.h"
#include "setup.h"

//***************************************************************************
// 
//***************************************************************************

static const char *VERSION        = "0.0.4";
static const char *DESCRIPTION    = "sedu ambi light control with data from softhddevice";
static const char *MAINMENUENTRY  = "Seduatmo";

//***************************************************************************
// Plugin
//***************************************************************************

class cPluginSeduatmo : public cPlugin 
{
   public:
      
      cPluginSeduatmo(void);
      virtual ~cPluginSeduatmo();
      virtual const char* Version(void)          { return VERSION; }
      virtual const char* Description(void)      { return DESCRIPTION; }
      virtual const char* CommandLineHelp(void)  { return 0; }
      virtual bool ProcessArgs(int argc, char* argv[]);
      virtual bool Initialize(void);
      virtual bool Start(void);
      virtual void Stop(void);
      virtual void Housekeeping(void)   { };
      virtual void MainThreadHook(void) { };
      virtual cString Active(void);
      virtual time_t WakeupTime(void);
      virtual const char* MainMenuEntry(void) { return cfg.showMainmenu ?  MAINMENUENTRY : 0; }
      virtual cOsdObject* MainMenuAction(void);
      virtual cMenuSetupPage* SetupMenu(void) { return new cSeduSetup; }
      virtual bool SetupParse(const char* Name, const char* Value);
      virtual bool Service(const char* Id, void* Data = NULL);
      virtual const char** SVDRPHelpPages(void);
      virtual cString SVDRPCommand(const char* Command, const char* Option, int &ReplyCode);

      int startAtmo();
      int stopAtmo();
      cSeduThread* update;

      int isRunning()
      {
         if (!update)
            return no;

         return update->isRunning();
      }

   private:

      cLedConfs ledConfs;
};

//***************************************************************************
// Plugins Main Menu
//***************************************************************************

class cSeduPluginMenu : public cMenuSetupPage
{
   public:

      cSeduPluginMenu(const char* title, cPluginSeduatmo* aPlugin);
      virtual ~cSeduPluginMenu() { };
      
      virtual eOSState ProcessKey(eKeys key);

   protected:

      void Store();
      cPluginSeduatmo* plugin;
      int effectSpeed;
};

cSeduPluginMenu::cSeduPluginMenu(const char* title, cPluginSeduatmo* aPlugin)
{
   SetTitle(title ? title : "");
   plugin = aPlugin;
   effectSpeed = cfg.effectSpeed;
   
   Clear();

   cOsdMenu::Add(new cMenuEditStraItem(tr("View Mode"), 
                                       (int*)&cfg.viewMode, 
                                       (int)cSeduService::vmCount,
                                       cSeduService::viewModes));

   Add(new cMenuEditIntItem(tr("Fixed Color Red"), &cfg.fixedR, 0, 255));
   Add(new cMenuEditIntItem(tr("Fixed Color Green"), &cfg.fixedG, 0, 255));
   Add(new cMenuEditIntItem(tr("Fixed Color Blue"), &cfg.fixedB, 0, 255));

   Add(new cMenuEditIntItem(tr("Effect Speed [ms]"), &effectSpeed, 100, 5000));

   SetHelp(0, 0, 0, 0);

   Display();
}

//***************************************************************************
// Process Key
//***************************************************************************

eOSState cSeduPluginMenu::ProcessKey(eKeys key)
{
   eOSState state = cOsdMenu::ProcessKey(key);
   
   if (key == kLeft || key == kRight)
   {
      if (cfg.viewMode == cSeduService::vmDetached && plugin->isRunning())
         plugin->stopAtmo();
      else if (cfg.viewMode != cSeduService::vmDetached && !plugin->isRunning())
         plugin->startAtmo();

      Display();
   }

   if (key == kOk)
   {
      cfg.effectSpeed = effectSpeed;
      Store();

      return osEnd;
   }

   return state;
}

void cSeduPluginMenu::Store()
{  
   plugin->SetupStore("FixedColorRed", cfg.fixedR);
   plugin->SetupStore("FixedColorGreen", cfg.fixedG);
   plugin->SetupStore("FixedColorBlue", cfg.fixedB);
   plugin->SetupStore("ViewMode", (int)cfg.viewMode);
   plugin->SetupStore("EffectSpeed", cfg.effectSpeed);

   Setup.Save();
}

//***************************************************************************
// Plugin 
//***************************************************************************

cPluginSeduatmo::cPluginSeduatmo(void)
{
   update = 0;
}

cPluginSeduatmo::~cPluginSeduatmo()
{
   stopAtmo();
}

bool cPluginSeduatmo::ProcessArgs(int argc, char* argv[])
{
   return true;
}

bool cPluginSeduatmo::Initialize(void)
{
   char* tmp;

   asprintf(&tmp, "%s/%s", cPlugin::ConfigDirectory("seduatmo"), "seduatmo.conf");

   if (!ledConfs.Load(tmp, true))
   {
      free(tmp);
      return fail;
   }

   free(tmp);

   cfg.createLeds(&ledConfs);

   return true;
}

bool cPluginSeduatmo::Start(void)
{
   startAtmo();

   return true;
}

int cPluginSeduatmo::startAtmo()
{
   if (update)
   {
      update->Stop();
      delete update;
   }

   update = new cSeduThread();

   if (cfg.grabWidth <= 0 || cfg.grabHeight <= 0 || !cfg.ledCount)
      tell(0, "Error: Invalid configuration in seduatmo.conf, aborting");
   else
      update->Start();

   return done;
}

int cPluginSeduatmo::stopAtmo()
{
   if (update)
      update->Stop();

   delete update;
   update = 0;

   return done;
}

void cPluginSeduatmo::Stop(void)
{
   stopAtmo();
}

cString cPluginSeduatmo::Active(void)
{
  return 0;
}

time_t cPluginSeduatmo::WakeupTime(void)
{
  return 0;
}

cOsdObject* cPluginSeduatmo::MainMenuAction(void)
{
   return new cSeduPluginMenu(MAINMENUENTRY, this);
}

bool cPluginSeduatmo::SetupParse(const char* Name, const char* Value)
{
   if      (!strcasecmp(Name, "LogLevel"))         cfg.loglevel = atoi(Value);
   else if (!strcasecmp(Name, "ShowMainmenu"))     cfg.showMainmenu = atoi(Value);
   else if (!strcasecmp(Name, "ViewMode"))         cfg.viewMode = (cSeduService::ViewMode)atoi(Value);

   else if (!strcasecmp(Name, "DetectCineBars"))   cfg.detectCineBars = (cSeduService::Cinebars)atoi(Value);

   else if (!strcasecmp(Name, "Threshold"))        cfg.threshold = atoi(Value);
   else if (!strcasecmp(Name, "Gamma"))            cfg.gamma = atoi(Value);
   else if (!strcasecmp(Name, "Frequence"))        cfg.frequence = atoi(Value);
   else if (!strcasecmp(Name, "Blacklevel"))       cfg.black = atoi(Value);
   else if (!strcasecmp(Name, "Smoothness"))       cfg.average = atoi(Value);

   else if (!strcasecmp(Name, "HorizontalDepth"))  cfg.xDeep = atoi(Value);
   else if (!strcasecmp(Name, "VerticalDepth"))    cfg.yDeep = atoi(Value);

   else if (!strcasecmp(Name, "LevelRed"))         cfg.adjRed = atoi(Value);
   else if (!strcasecmp(Name, "LevelGreen"))       cfg.adjGreen = atoi(Value);
   else if (!strcasecmp(Name, "LevelBlue"))        cfg.adjBlue = atoi(Value);

   else if (!strcasecmp(Name, "FixedColorRed"))    cfg.fixedR = atoi(Value);
   else if (!strcasecmp(Name, "FixedColorGreen"))  cfg.fixedG = atoi(Value);
   else if (!strcasecmp(Name, "FixedColorBlue"))   cfg.fixedB = atoi(Value);
   else if (!strcasecmp(Name, "EffectSpeed"))      cfg.effectSpeed = atoi(Value);

   else if (!strcasecmp(Name, "SeduMode"))         cfg.seduMode = (cSeduService::SeduMode)atoi(Value);
   else if (!strcasecmp(Name, "SeduRGBOrder"))     strcpy(cfg.seduRGBOrder, Value);

   else
      return false;

   return true;
}

bool cPluginSeduatmo::Service(const char* Id, void* Data)
{
   return false;
}

cString cPluginSeduatmo::SVDRPCommand(const char* Command, const char* Option, int &ReplyCode)
{
   if (!strcasecmp(Command, "MODE")) 
   {
      if (Option && strcasecmp(Option, "atmo") == 0) 
      {
         cfg.viewMode = cSeduService::vmAtmo;
         startAtmo();
         ReplyCode = 550;
         return "atmo mode activated";
      }
      else if (Option && strcasecmp(Option, "fixed") == 0) 
      {
         cfg.viewMode = cSeduService::vmFixedCol;
         startAtmo();
         ReplyCode = 550;
         return "fixed color activated";
      }
      else if (Option && strcasecmp(Option, "rainbow") == 0) 
      {
         cfg.viewMode = cSeduService::vmRainbow;
         startAtmo();
         ReplyCode = 550;
         return "rainbow effect activated";
      }
      else if (Option && strcasecmp(Option, "wheel") == 0) 
      {
         cfg.viewMode = cSeduService::vmColorWheel;
         startAtmo();
         ReplyCode = 550;
         return "color wheel effect activated";
      }
      else if (Option && strcasecmp(Option, "wheelstatic") == 0) 
      {
         cfg.viewMode = cSeduService::vmColorWheelStatic;
         startAtmo();
         ReplyCode = 550;
         return "static color wheel activated";
      }
      else if (Option && strcasecmp(Option, "black") == 0) 
      {
         cfg.viewMode = cSeduService::vmBlack;
         startAtmo();
         
         ReplyCode = 550;
         return "stripes black";
      }
      else if (Option && strcasecmp(Option, "detach") == 0) 
      {
         cfg.viewMode = cSeduService::vmDetached;
         stopAtmo();
         
         ReplyCode = 550;
         return "stripes detached";
      }
      else
      {
         ReplyCode = 901;
         return "Error: Unexpected option";
      }
   }

   return 0;
}

const char** cPluginSeduatmo::SVDRPHelpPages(void) 
{ 
   static const char* HelpPages[] = 
   {
      "MODE <mode>\n"
      "    Set mode {atmo|fixed|rainbow|wheel|wheelstatic|black|detach}\n",
      0
   };

   return HelpPages;
}

//***************************************************************************
// Class Setup
//***************************************************************************

const char* cSeduSetup::seduRGBOrders[] = 
{
   "RGB",
   "RBG",
   "BGR",
   "BRG",
   "GBR",
   "GRB",

   0
};

int cSeduSetup::toOrderIndex(const char* order)
{
   for (int i = 0; seduRGBOrders[i];i++)
      if (strcasecmp(seduRGBOrders[i], order) == 0)
         return i;

   return na;
}

//***************************************************************************
// Object
//***************************************************************************

cSeduSetup::cSeduSetup()
{
   data = cfg;
   data.copyLeds(&cfg);

   // sedu mode

   seduModes[smMiniDMX] 	= "miniDMX_512";
   seduModes[smTpm2] 		= "tpm2";
   seduModes[smMiniDMX_A1] 	= "miniDMX_192";
   seduModes[smMiniDMX_B0] 	= "miniDMX_768";

   // rgb order

   rgbOrderIndex = toOrderIndex(data.seduRGBOrder);

   if (rgbOrderIndex = na)
      rgbOrderIndex = 0;

   // Cinema Bars

   cineBars[0] = "Horizontal";
   cineBars[1] = "Vertical";
   cineBars[2] = "Both";

   Setup();
}

//***************************************************************************
// Setup
//***************************************************************************

void cSeduSetup::Setup() 
{
   Clear();

   Add(new cMenuEditIntItem(tr("Log level"), &data.loglevel, 0, 3));
   Add(new cMenuEditBoolItem(tr("Show mainmenu"), &data.showMainmenu));

   Add(new cMenuEditIntItem(tr("Frequence [Hz]"), &data.frequence, 1, 100));

   Add(new cMenuEditStraItem(tr("Detect cinema bars"), (int*)&data.detectCineBars, 3, cineBars));

   Add(new cMenuEditIntItem(tr("Threshold"), &data.threshold, 0, 255));
   Add(new cMenuEditIntItem(tr("Gamma (10-30)"), &data.gamma, 10, 30));
   Add(new cMenuEditIntItem(tr("Horizontal depth"), &data.xDeep, 0, 5));
   Add(new cMenuEditIntItem(tr("Vertical depth"), &data.yDeep, 0, 5));
   Add(new cMenuEditIntItem(tr("Smoothness"), (int*)&data.average, 1, 100));
   Add(new cMenuEditIntItem(tr("Blacklevel"), &data.black, 0, 255));

   Add(new cMenuEditIntItem(tr("Level red [%]"), &data.adjRed, 0, 100));
   Add(new cMenuEditIntItem(tr("Level green [%]"), &data.adjGreen, 0, 100));
   Add(new cMenuEditIntItem(tr("Level blue [%]"), &data.adjBlue, 0, 100));

   Add(new cMenuEditStraItem(tr("SEDU mode"), (int*)&data.seduMode, 4, seduModes));
   Add(new cMenuEditStraItem(tr("SEDU RGB order"), &rgbOrderIndex, 6, seduRGBOrders));
}

eOSState cSeduSetup::ProcessKey(eKeys Key) 
{
   eOSState state = cMenuSetupPage::ProcessKey(Key);

   // nothing yet ..

   return state;
}

void cSeduSetup::Store()
{
   strcpy(data.seduRGBOrder, seduRGBOrders[rgbOrderIndex]);

   cfg = data;
   cfg.copyLeds(&data);

   SetupStore("LogLevel", data.loglevel);
   SetupStore("ShowMainmenu", data.showMainmenu);
   SetupStore("ViewMode", (int)data.viewMode);

   SetupStore("DetectCineBars", data.detectCineBars);

   SetupStore("Frequence", data.frequence);
   SetupStore("Threshold", data.threshold);
   SetupStore("Gamma", data.gamma);
   SetupStore("Blacklevel", data.black);
   SetupStore("Smoothness", data.average);

   SetupStore("HorizontalDepth", data.xDeep);
   SetupStore("VerticalDepth", data.yDeep);

   SetupStore("LevelRed", data.adjRed);
   SetupStore("LevelGreen", data.adjGreen);
   SetupStore("LevelBlue", data.adjBlue);

   SetupStore("FixedColorRed", data.fixedR);
   SetupStore("FixedColorGreen", data.fixedG);
   SetupStore("FixedColorBlue", data.fixedB);
   SetupStore("EffectSpeed", data.effectSpeed);

   SetupStore("SeduMode", data.seduMode);
   SetupStore("SeduRgbOrder", data.seduRGBOrder);
}

//***************************************************************************
// VDR Internal
//***************************************************************************

VDRPLUGINCREATOR(cPluginSeduatmo);
